feat(platform/github): flag to control whether PRs can be edited by maintainers if `forkToken`is set

devversion opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

A new option called `forkModeAllowMaintainerEdits` is introduced. This option is only relevant if Renovate is used with `forkToken`.
* The option is by default `true`, keeping backwards compatibilty
* The option will only have an effect for `github`. 

The option decides whether maintainer edits are allowed to edit pull requests.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

If a maintainer pushes changes to a fork renovate PR, the branch will never be deleted by Renovate because there are changes not corresponding to the Git account configured in Renovate.

This prevents future updates as Renovate basically locks down the branch and marks it as `pr-edited`, preventing it from being overridden or cleaned-up as part of the scheduled branch deletion.

We can fix this by encouraging that maintainers cherry-pick the commit and make changes in a separate PR. Users of the experimental fork mode option can do this by setting `prAllowMaintainerEdits = false`. It is not possible for maintainers to trivially delete the branch of the e.g. fork robot account because they would need to log into that and not every team member would e.g. necessarily have access to the robot account.

- Fixes #16657

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [X] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

rarkins commented on lib/modules/platform/github/index.ts:
```diff
@@ -1494,7 +1494,10 @@ export async function createPr({
   // istanbul ignore if
   if (config.forkToken) {
     options.token = config.forkToken;
-    options.body.maintainer_can_modify = true;
```
It had been my assumption until now that this feature was quite useful for users. Therefore I want to think whether this change is necessary or if there's "another way".

devversion replied:
yeah, it totally is. I think in the issue I was advocating for a simpler branch deletion policy, like:

* If a PR is for the onboarding branch, never delete it
* If a PR is modified
   * only delete if there is no associated PR

but it seemed like changing the process here would be a rather larger undertaking, while for forked PRs this is already an improvement compared to renovate suddenly not updating a certain dependency at all anymore

viceice replied:
this is a breaking change, so it should better be configurable and enabled by default

devversion replied:
yeah, I can make it configurable. Though could also see it as a fix because the current "intended behavior" does not work and has more significant consequences (like people not knowing that they miss updates after having edited a fork branch)

viceice replied:
doesn't the delete branch after merge function (GitHub repo settings) solve the issue too?

devversion replied:
Interesting idea! (didn't know this existed). It looks like it wouldn't be able to delete branches from the fork account though. e.g. consider https://github.com/angular/dev-infra/pull/821 where the PR branch is not an upstream branch
<img width="778" alt="Screenshot 2023-01-10 at 21 23 35" src="https://user-images.githubusercontent.com/4987015/211654685-2e49f8e8-887f-4a94-9de5-06e813c903a2.png">

We had the setting enabled but it didn't delete the branch. Likely a maintainer cannot delete a branch from a user fork

rarkins replied:
I agree with https://github.com/renovatebot/renovate/pull/19771#discussion_r1066218416

viceice reviewed (changes requested):
needs to be configurable and enabled by default to be not breaking

rarkins reviewed (changes requested):
1. This is only relevant to forkingRenovate. Right now it's configurable for both and the docs don't make it clear enough. Let's try to avoid confusing the 99% of users not affected by this
2. Can this be made global config instead (i.e. controlled by the bot admin, not the repo admin)?

HonkingGoose commented on lib/config/options/index.ts:
```diff
@@ -1555,6 +1555,14 @@ const options: RenovateOptions[] = [
+  {
+    name: 'prAllowMaintainerEdits',
+    description:
+      'Whether pull requests created should allow for maintainer edits.',
```
```suggestion
      'Decides if maintainers can edit Renovate pull requests.',
```

How about something like this?

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -2179,6 +2179,18 @@ Defaults to `update`, but can also be set to `branch`.
 
+## prAllowMaintainerEdits
+
+This setting controls whether pull requests created by Renovate will allow edits by maintainers.
```
```suggestion
Use `prAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests.
```

How about something like this?

devversion replied:
thanks!

devversion commented:
I've addressed all feedback (and appreciate anybody having another look. Thanks for all the comments)

<sup>Github seems to just send it to one of the reviewers</sup>

viceice commented on lib/config/options/index.ts:
```diff
@@ -1555,6 +1555,14 @@ const options: RenovateOptions[] = [
+    type: 'boolean',
+    supportedPlatforms: ['github'],
+    default: true,
+    globalOnly: true,
```
so only the bot owner can change it? how would this PR help with forking renovate app?

devversion replied:
good point. should we go back to make it a repository setting?

devversion replied:
I've switched it back to a per-repo setting since the forking renovate GitHub app is a good argument. Since you also want this to be configurable, this seems like the only option.

viceice commented on lib/modules/platform/github/index.ts:
```diff
@@ -1494,12 +1494,12 @@ export async function createPr({
       base,
       body,
       draft: draftPR,
+      maintainer_can_modify: !!GlobalConfig.get('prAllowMaintainerEdits'),
```
this should be moved back to the if condition below, as it only works with forkToken.

devversion commented:
I've incorporated all the docs suggestions and switched back to a repo config setting. A global option would simplify this a bit, but would make it impossible to use this option when e.g. the forking renovate GitHub app is used.

Sorry. I've also force-pushed to clean-up the commits which made it more difficult to review- a full review is easier in this situation anyway (and very likely needed anyway)

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -2181,6 +2181,17 @@ Defaults to `update`, but can also be set to `branch`.
+
+Use `prAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests.
+
+When Renovate runs in a fork, this allows project maintainers to make manual changes to the
+Renovate PR branch, without needing to create another new PR.
```
```suggestion
When Renovate runs in a fork, this allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.
```

devversion replied:
updated (sorry was OOO until today)

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -2181,6 +2181,16 @@ Defaults to `update`, but can also be set to `branch`.
 This sets the level the postUpgradeTask runs on, if set to `update` the postUpgradeTask will be executed for every dependency on the branch.
 If set to `branch` the postUpgradeTask is executed for the whole branch.
 
+## prAllowMaintainerEdits
+
+Use `prAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests.
+
+When Renovate runs in a fork, this allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.
+
+<!-- prettier-ignore -->
+!!! note
+    This option is only relevant if you set `forkToken`.
```
```suggestion
## forkModeAllowMaintainerEdits

Use `forkModeAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests when in fork mode.

When Renovate runs in a fork, this allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.

<!-- prettier-ignore -->
!!! note
    This option is only relevant if you set `forkToken`.
```


This will require reordering within the file too

rarkins commented on lib/config/options/index.ts:
```diff
@@ -1555,6 +1555,13 @@ const options: RenovateOptions[] = [
     default: 'strict',
   },
+  {
+    name: 'prAllowMaintainerEdits',
```
```suggestion
    name: 'forkModeAllowMaintainerEdits',
```

rarkins commented on lib/config/types.ts:
```diff
@@ -230,6 +230,7 @@ export interface RenovateConfig
   postUpdateOptions?: string[];
   prConcurrentLimit?: number;
   prHourlyLimit?: number;
+  prAllowMaintainerEdits?: boolean;
```
```suggestion
  forkModeAllowMaintainerEdits?: boolean;
```

rarkins commented on lib/workers/repository/config-migration/pr/index.ts:
```diff
@@ -100,6 +100,7 @@ ${
         sourceBranch: branchName,
         // TODO #7154
         targetBranch: config.defaultBranch!,
+        allowMaintainerEdits: config.prAllowMaintainerEdits,
```
```suggestion
        allowMaintainerEdits: config.forkModeAllowMaintainerEdits,
```

devversion commented on lib/config/options/index.ts:
```diff
@@ -1555,6 +1555,13 @@ const options: RenovateOptions[] = [
+    description: 'Decides if maintainers can edit Renovate pull requests.',
+    type: 'boolean',
+    supportedPlatforms: ['github'],
+    default: true,
```
@viceice brought up a good point on this. Users of e.g. the Renovate forking app would not be able to benefit from this option then.

devversion commented:
I'm good restricting the option name also to `forkMode`. Updated as per review suggestions. One extra note: I changed the `createPr` option from `allowMaintainerEdits` also to `forkModeAllowMaintainerEdits` because otherwise it would be quite ambiguous. We could even go further and move it into platform options I assume.

devversion commented:
FYI: I will wait with any docs changes until we have settled on runtime code, naming, or if this should be a global only flag etc.

devversion commented on docs/usage/configuration-options.md:
```diff
@@ -875,6 +875,16 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
+
+<!-- prettier-ignore -->
+!!! note
+    This option is only relevant if you set `forkToken`.
```
The `a` is intentionally omitted because it should refer to the `forkToken` option. I think adding `a` would make this a little less clear.

HonkingGoose replied:
It felt a bit wrong to me. 😄 I'll defer to the judgment of a native English speaker on this. 😉

devversion replied:
me too 😄

viceice commented on lib/workers/repository/onboarding/pr/index.ts:
```diff
@@ -174,6 +174,7 @@ If you need any further assistance then you can also [request help here](${
       const pr = await platform.createPr({
         sourceBranch: config.onboardingBranch!,
         targetBranch: config.defaultBranch!,
+        forkModeAllowMaintainerEdits: config.forkModeAllowMaintainerEdits,
```
@rarkins i think we should pass this via the platform options below?

devversion replied:
yeah, I was wondering too (https://github.com/renovatebot/renovate/pull/19771#issuecomment-1426039302). 

will make the change based on what you decide

devversion replied:
friendly ping so I can make all the suggested changes

rarkins replied:
Yes, please pass it as a platform option

devversion commented on docs/usage/configuration-options.md:
```diff
@@ -875,6 +875,16 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
+
+Use `forkModeAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests when in fork mode.
+
+This allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.
```
I've tried to add more info here, but it's a little odd because the old workflow is basically *broken* 😄

viceice commented on docs/usage/configuration-options.md:
```diff
@@ -894,6 +894,20 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
+
+If GitHub pull requests are created from a [fork repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo), the PR author can decide to allow upstream repository to modify the PR directly.
+
+Allowing maintainers to edit pull requests directly is helpful when Renovate pull requests require additional changes. The reviewer can simply push to the pull request without having to create a new PR. [More details here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
```
```suggestion
Allowing maintainers to edit pull requests directly is helpful when Renovate pull requests require additional changes.
The reviewer can simply push to the pull request without having to create a new PR. [More details here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
```
one sentence per line

viceice commented on docs/usage/configuration-options.md:
```diff
@@ -894,6 +894,20 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
+
+Allowing maintainers to edit pull requests directly is helpful when Renovate pull requests require additional changes. The reviewer can simply push to the pull request without having to create a new PR. [More details here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
+
+You may decide to disallow edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657). If this option is enabled, reviewers will need to create a new PR if additional changes are needed.
```
```suggestion
You may decide to disallow edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657).
If this option is enabled, reviewers will need to create a new PR if additional changes are needed.
```

viceice commented on lib/config/options/index.ts:
```diff
@@ -1929,6 +1929,14 @@ const options: RenovateOptions[] = [
+      'Disallow maintainers to push to Renovate pull requests when running in fork mode.',
+    type: 'boolean',
+    supportedPlatforms: ['github'],
+    default: true,
```
```suggestion
    default: false,
```
so we don't break existing users

devversion replied:
good catch. I reversed the option with the recent changs

viceice commented on lib/modules/platform/types.ts:
```diff
@@ -92,6 +92,7 @@ export type PlatformPrOptions = {
   bbUseDefaultReviewers?: boolean;
   gitLabIgnoreApprovals?: boolean;
   usePlatformAutomerge?: boolean;
+  githubForkModeDisallowMaintainerEdits?: boolean;
```
```suggestion
  forkModeDisallowMaintainerEdits?: boolean;
```
use same name as config option

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -894,6 +894,22 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
 For example: you're following the `next` tag, but later the stream you actually want is called `stable` instead.
 If `next` is no longer getting updates, you must switch your `followTag` to `stable` to get updates again.
 
+## forkModeDisallowMaintainerEdits
```
```suggestion
## forkModeBlockMaintainerEdits
```

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -894,6 +894,22 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
 
+## forkModeDisallowMaintainerEdits
+
+Use `forkModeDisallowMaintainerEdits` to disallow maintainers from editing Renovate's pull requests when in fork mode.
```
```suggestion
Use `forkModeBlockMaintainerEdits` to block maintainers from editing Renovate's pull requests when in fork mode.
```

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -894,6 +894,22 @@ Renovate follows tags _strictly_, this can cause problems when a tagged stream i
+Allowing maintainers to edit pull requests directly is helpful when Renovate pull requests require additional changes.
+The reviewer can simply push to the pull request without having to create a new PR. [More details here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
+
+You may decide to disallow edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657).
```
```suggestion
You may decide to block edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657).
```

HonkingGoose commented on lib/config/options/index.ts:
```diff
@@ -1929,6 +1929,14 @@ const options: RenovateOptions[] = [
     supportedPlatforms: ['gitlab'],
   },
+  {
+    name: 'forkModeDisallowMaintainerEdits',
```
```suggestion
    name: 'forkModeBlockMaintainerEdits',
```

HonkingGoose commented on lib/config/options/index.ts:
```diff
@@ -1929,6 +1929,14 @@ const options: RenovateOptions[] = [
+  {
+    name: 'forkModeDisallowMaintainerEdits',
+    description:
+      'Disallow maintainers to push to Renovate pull requests when running in fork mode.',
```
```suggestion
      'Block maintainer pushes to Renovate pull requests when running in fork mode.',
```

HonkingGoose commented on lib/config/types.ts:
```diff
@@ -230,6 +230,7 @@ export interface RenovateConfig
   postUpdateOptions?: string[];
   prConcurrentLimit?: number;
   prHourlyLimit?: number;
+  forkModeDisallowMaintainerEdits?: boolean;
```
```suggestion
  forkModeBlockMaintainerEdits?: boolean;
```

HonkingGoose commented on lib/modules/platform/github/index.spec.ts:
```diff
@@ -2180,6 +2180,96 @@ describe('modules/platform/github/index', () => {
+          prBody: 'PR can be edited by maintainers.',
+          labels: null,
+          platformOptions: {
+            forkModeDisallowMaintainerEdits: false,
```
```suggestion
            forkModeBlockMaintainerEdits: false,
```

HonkingGoose commented on lib/modules/platform/github/index.spec.ts:
```diff
@@ -2180,6 +2180,96 @@ describe('modules/platform/github/index', () => {
+          prBody: 'PR *cannot* be edited by maintainers.',
+          labels: null,
+          platformOptions: {
+            forkModeDisallowMaintainerEdits: true,
```
```suggestion
            forkModeBlockMaintainerEdits: true,
```

HonkingGoose commented on lib/modules/platform/github/index.ts:
```diff
@@ -1499,7 +1499,8 @@ export async function createPr({
     options.token = config.forkToken;
-    options.body.maintainer_can_modify = true;
+    options.body.maintainer_can_modify =
+      platformOptions?.forkModeDisallowMaintainerEdits !== true;
```
```suggestion
      platformOptions?.forkModeBlockMaintainerEdits !== true;
```

HonkingGoose commented on lib/modules/platform/types.ts:
```diff
@@ -92,6 +92,7 @@ export type PlatformPrOptions = {
   bbUseDefaultReviewers?: boolean;
   gitLabIgnoreApprovals?: boolean;
   usePlatformAutomerge?: boolean;
+  forkModeDisallowMaintainerEdits?: boolean;
```
```suggestion
  forkModeBlockMaintainerEdits?: boolean;
```

HonkingGoose commented on lib/workers/repository/update/pr/index.ts:
```diff
@@ -50,6 +50,7 @@ export function getPlatformPrOptions(
     azureWorkItemId: config.azureWorkItemId,
     bbUseDefaultReviewers: config.bbUseDefaultReviewers,
     gitLabIgnoreApprovals: config.gitLabIgnoreApprovals,
+    forkModeDisallowMaintainerEdits: config.forkModeDisallowMaintainerEdits,
```
```suggestion
    forkModeBlockMaintainerEdits: config.forkModeBlockMaintainerEdits,
```

HonkingGoose reviewed (commented):
How about using a simpler name like `forkModeBlockMaintainerEdits`?

The word `disallow` feels clunky to me. I think a word like `block` is easier to understand.

I'll let the maintainers decide if they want to apply my suggestions.

devversion commented:
Yeah, if the maintainers think they prefer `block`, I can make the change. I personally prefer `(dis)allow` since that is the terminology used by GitHub as well. Blocking feels different to me (but I'm a non-native speaker) See:

<img width="319" alt="image" src="https://user-images.githubusercontent.com/4987015/220593231-68c5a045-d045-4286-99bc-524726493a48.png">

and https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork.

viceice commented:
i think disallow is the right thing 🤷‍♂️

HonkingGoose commented:
> I personally prefer `(dis)allow` since that is the terminology used by GitHub as well.

Good point, let's stick with `dis(allow)` then. 😄

HonkingGoose commented:
@viceice can you resolve my PR review comments? I can't do this myself: [^docs]

> You can resolve a conversation in a pull request if you opened the pull request or if you have write access to the repository where the pull request was opened.

[^docs]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#resolving-conversations
